### DISCUSSION TITLE
Building labwc on NixOS / with Nix

### DISCUSSION THREAD

**Discussion by datMaffin** [labwc/labwc#1338] [Show and tell]
Add the following `shell.nix` file to the root of the labwc repo directory:

```nix
{ pkgs ? import <nixpkgs> {} }:
  pkgs.mkShell {
    nativeBuildInputs = [
      # new wlroots dependencies for seat feature, which is required by labwc
      pkgs.seatd
      # new wlroots dependencies for DRM feature, which is used in labwc
      pkgs.hwdata
      pkgs.libdisplay-info
      pkgs.libliftoff
    ];
    inputsFrom = [
      pkgs.labwc
      pkgs.wlroots
    ];
}
```

Then run `nix-shell` in that directory. For a successful build use
```
meson setup --buildtype plain build/
meson compile -C build/
```

The `--buildtype plain` is needed in NixOs such that fortify source is not causing the compile to error-out.

---

*This nix shell was tested with the current 23.11 nixpkgs.*

---

What this script does is basically install all the dependencies for the current labwc (and wlroots) from nixpkgs (the `inputsFrom`). 

But because master labwc needs a more up-to-date wlroots than what nixpkgs offers, we need to build wlroots ourselves. Luckily the build process is automatically done by Meson, so the only thing that needs to be done is to add the build dependencies for the upstream release of wlroots. (The additional `nativeBuildInputs` I added.)

**Comment by datMaffin** [comment]
With 24.05 nixpkgs, there is currently no need to specify any additional dependencies.

I.e., a `shell.nix` file with the content
```nix
{ pkgs ? import <nixpkgs> {} }:
  pkgs.mkShell {
    inputsFrom = [
      pkgs.labwc
    ];
}
```
is sufficient.

I will leave the original post as-is. Such that there is an example for the future, for when additional dependencies or a different wlroots being needed.

**Comment by DogLooksGood** [comment]
It seems that the keyboard layout settings doesn't work. It can't find any layout except `us` (qwerty).

**Reply by johanmalm** [reply]
Are you referring to the layout chooser in `labwc-tweaks`, or describing `XKB_DEFAULT_LAYOUT=whatever` not working with `labwc`, or something else? 

Ref:
-  https://github.com/labwc/labwc/blob/c2928027becba26f058b544d4bd8c72895c8d7cf/docs/environment#L20

**Reply by datMaffin** [reply]
> It seems that the keyboard layout settings doesn't work. It can't find any layout except `us` (qwerty).

I quickly compiled the latest master, and my config (which is setting up my keyboard layout) works as expected.

My config is:
```nix
    "xdg/labwc/environment".text = ''
      MOZ_ENABLE_WAYLAND=1
      XKB_DEFAULT_LAYOUT=de-lsgtmod
      _JAVA_AWT_WM_NONREPARENTING=1
    '';
```

de-lsgtmod is my (custom) layout.

**Reply by DogLooksGood** [reply]
You're right! I just realized by XBD_KEYBOARD_LAYOUT format is wrong. I'm using dvp, and it should be us(dvp) instead of just dvp.

**Reply by datMaffin** [reply]
Also, the environment variable seems to have a different name than what you wrote:
```diff
- XBD_KEYBOARD_LAYOUT
+ XKB_DEFAULT_LAYOUT
```
